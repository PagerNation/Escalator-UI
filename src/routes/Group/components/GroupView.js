import React from "react";
import "./GroupView.scss";
import { Divider, Label, Grid, Header, Button, Confirm, Icon, Segment, Card, Feed } from 'semantic-ui-react';
import _ from 'lodash';
import InlineEditable from '../../../components/shared/InlineEditable';
import classNames from 'classnames';
import RemoveSubscriberModal from './RemoveSubscriberModal';
import TicketView from './TicketView';

class GroupView extends React.Component {

  constructor() {
    super();
    this.state = {
      confirmOpen: false,
      showOpenTickets: true,
      selectedOnCall: [],
      selectedBenched: []
    };
    _.bindAll(this,
      "toggleConfirm",
      "handleLeave",
      "toggleSelectOnCall",
      "toggleSelectBenched",
      "handleRemoveSubscribers",
      "handleAddSubscribers",
      "handleProcessRequest",
      "handleEditPagingInterval",
      "toggleRemoveModal",
      "handleTicketAcknowledgement",
      "handleScheduleSubscribers"
    );
  }

  toggleRemoveModal() {
    const subs = _.extend({}, this.props.group.escalationPolicy.subscribers);
    const subsToRemove = [];
    this.state.selectedOnCall.forEach((i) => {
      subsToRemove.push(subs[i]);
    });
    _.forEach(subsToRemove, (sub) => {
      sub.name = _.find(this.props.group.users, (user) => user._id === sub.user).name;
    });
    this.refs["removeSubModal"].open(subsToRemove);

  }

  toggleSelectOnCall(index) {
    if (this.props.group.admins.includes(this.props.user._id)) {
      this.setState({
        selectedOnCall: _.xor(this.state.selectedOnCall, [index])
      });
    }
  }

  toggleSelectBenched(index) {
    if (this.props.group.admins.includes(this.props.user._id)) {
      this.setState({
        selectedBenched: _.xor(this.state.selectedBenched, [index])
      });
    }
  }

  toggleConfirm() {
    this.setState({
      confirmOpen: !this.state.confirmOpen
    });
  }

  handleLeave() {
    this.props.leaveGroup(this.props.group.name, this.props.user._id).then(this.toggleConfirm);
  }

  handleRemoveSubscribers(subsToRemove) {
    if (!subsToRemove.length) return;

    const ep = {};
    let subs = _.differenceBy(this.props.group.escalationPolicy.subscribers, subsToRemove, (user) => user._id);
    subs = _.map(subs, (sub) => _.omit(sub, "_id", "name"));
    _.extend(ep, this.props.group.escalationPolicy, {subscribers: subs});
    this.props.updateEscalationPolicy(this.props.group.name, _.omit(ep, "_id"));
    this.setState({
      selectedOnCall: []
    });
  }

  handleScheduleSubscribers(subsToSchedule) {
    if (!subsToSchedule.length) return;

    let subs = _.extend([], this.props.group.escalationPolicy.subscribers);
    _.forEach(subsToSchedule, (sub) => {
      const match = _.find(subs, (s) => s.user === sub.user);
      match.deactivateDate = sub.deactivateDate;
      match.reactivateDate = sub.reactivateDate;
    });
    subs = _.map(subs, (sub) => _.omit(sub, "_id", "name"));
    const ep = _.extend({}, this.props.group.escalationPolicy, {subscribers: subs});
    this.props.updateEscalationPolicy(this.props.group.name, _.omit(ep, "_id"));
  }

  handleAddSubscribers() {
    const subs = this.props.group.escalationPolicy.subscribers.map((user) => _.omit(user, "_id", "name"));
    const subIds = subs.map((user) => user.user);
    const benched = [,...this.props.group.users].filter((user) =>
      subIds.indexOf(user._id) === -1
    );
    this.state.selectedBenched.forEach((i) => subs.push({
      user: benched[i]._id,
      active: true,
      deactivateDate: null,
      reactivateDate: null
    }));
    const ep = {};
    _.extend(ep, this.props.group.escalationPolicy, {subscribers: subs});
    this.props.updateEscalationPolicy(this.props.group.name, _.omit(ep, "_id"));
    this.setState({
      selectedBenched: []
    });
  }

  handleEditPagingInterval(values) {
    const ep = {};
    _.extend(ep, this.props.group.escalationPolicy, {pagingIntervalInMinutes: values.pagingIntervalInMinutes});
    this.props.updateEscalationPolicy(this.props.group.name, _.omit(ep, "_id"));
  }

  handleProcessRequest(userId, approved) {
    this.props.processRequest(this.props.group.name, userId, approved);
  }

  handleTicketAcknowledgement(ticketId) {
    this.props.acknowledgeTicket(ticketId);
  }

  componentWillMount() {
    this.props.fetchGroup(this.props.params.groupId);
    this.props.fetchGroupTickets({
      groupNames: this.props.params.groupId
    });
    this.props.fetchOpenGroupTickets({
      groupNames: this.props.params.groupId
    });
  }

  active() {
    const subs = [...this.props.group.escalationPolicy.subscribers];
    _.forEach(subs, (sub) => {
      sub.name = _.find(this.props.group.users, (user) => user._id === sub.user).name;
    });
    return subs.map((user, i) =>
      <a
        className={classNames("box arrow_box", {"selected": this.state.selectedOnCall.includes(i)})}
        onClick={() => this.toggleSelectOnCall(i)}
        key={i}>
        {user.name}
      </a>
    );
  };

  benched() {
   const subIds = this.props.group.escalationPolicy.subscribers.map(u => u.user);
   return [,...this.props.group.users].filter((user) =>
     subIds.indexOf(user._id) === -1
   ).map((user, i) =>
      <a
        className={classNames("box", {"selected": this.state.selectedBenched.includes(i)})}
        onClick={() => this.toggleSelectBenched(i)}
        key={i}>
        {user.name}
      </a>
    )
  };

  onCall() {
    const group = this.props.group;
    if (group.escalationPolicy.subscribers.length) {
      const uid = group.escalationPolicy.subscribers[0].user;
      const user = group.users.filter(u => u._id === uid)[0];
      return <Header as="h4">User on call:{this.userLink(user)}</Header>;
    }
  };

  userLink(user) {
    return user ? <a href={'/user/' + user._id +'/'}>{user.name}</a> : <span>No one</span>;
  };

  escalationInterval() {
    return (
      <span>
        Escalation Interval: <Label color='teal' horizontal>{this.props.group.escalationPolicy.pagingIntervalInMinutes}</Label> minutes
      </span>
    );
  };

  escalationIntervalAdmin() {
    return (
      <span>
        Escalation Interval:<Label color='teal' horizontal>
          <InlineEditable
            name="pagingIntervalInMinutes"
            value={this.props.group.escalationPolicy.pagingIntervalInMinutes}
            onChange={this.handleEditPagingInterval} />
        </Label>
        minutes
      </span>
    );
  };

  renderAdmin() {
    return this.props.group.joinRequests
          && this.props.group.joinRequests.length > 0
          && this.props.group.admins.includes(this.props.user._id) && (
      <div>
        <Divider/>
        <Header as="h3">Administration</Header>
        <Grid className="group-requests">
          <Grid.Column mobile={16} computer={8}>
            <Header as="h4">Pending membership requests:</Header>
            <div>
              {this.props.group.joinRequests.map((user, i) =>
                <Segment key={i} raised>
                  <Header as="h5">{user.name}</Header>
                  <span>{user.email}</span>
                  <Button style={{marginTop: "-20px"}} floated="right" color="green" onClick={() => this.handleProcessRequest(user._id, true)}>Approve</Button>
                  <Button style={{marginTop: "-20px"}} className="section-btn" floated="right" color="red" onClick={() => this.handleProcessRequest(user._id, false)}>Deny</Button>
                </Segment>
              )}
            </div>
          </Grid.Column>
        </Grid>
      </div>
    );
  }

  renderEPControls() {
    return this.props.group.admins.includes(this.props.user._id) && (
      <div>
        <Button
          className="move-btn"
          size="mini"
          disabled={this.state.selectedBenched.length == 0}
          onClick={this.handleAddSubscribers}
        >
          <Icon name="chevron left" />
        </Button>
        <Button
          className="move-btn"
          size="mini"
          disabled={this.state.selectedOnCall.length == 0}
          onClick={this.toggleRemoveModal}
        >
          <Icon name="chevron right" />
        </Button>
      </div>
    );
  }

  render() {
    const leaveButton = this.props.group && _.find(this.props.group.users, (user) => user._id === this.props.user._id) && (
      <Button className="action-button" onClick={this.toggleConfirm}>Leave Group</Button>
    );

    const confirm = (
      <Confirm
        open={this.state.confirmOpen}
        content={`Are you sure you want to leave this group?`}
        confirmButton="Leave"
        onCancel={this.toggleConfirm}
        onConfirm={this.handleLeave}
      />
    );

    return this.props.group && (
      <div>
        {confirm}
        <RemoveSubscriberModal
          ref="removeSubModal"
          onRemove={this.handleRemoveSubscribers}
          onSchedule={this.handleScheduleSubscribers}
        />
        <Header as="h1">{this.props.group.name}</Header>
        {leaveButton}
        {this.onCall()}
        {this.props.group.admins.includes(this.props.user._id) ?
          this.escalationIntervalAdmin() :
          this.escalationInterval()
        }
        <Divider/>
        <Grid>
          <Grid.Column mobile={16} computer={7}>
            <h3>Escalation Order:</h3>
            {this.active()}
          </Grid.Column>
          <Grid.Column verticalAlign="middle" mobile={16} computer={1}>
            {this.renderEPControls()}
          </Grid.Column>
          <Grid.Column mobile={16} computer={7}>
            <h3>Not On Call:</h3>
            {this.benched()}
          </Grid.Column>
        </Grid>
        {this.renderAdmin()}
        <TicketView
          user={this.props.user}
          tickets={this.props.tickets}
          openTickets={this.props.openTickets}
          handleTicketAcknowledgement={this.handleTicketAcknowledgement}
        />
      </div>
    );
  }
}

export default GroupView;
