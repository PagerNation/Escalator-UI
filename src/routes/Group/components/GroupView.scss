.on-call {
  color: green !important;
  font-weight: 800;
}

button.ui.button.action-button {
  float: right;
  clear: both;
}

.box {
  border: 3px solid #c2e1f5;
  background: #88b7d5;
  display: block;
  position: relative;
  padding: 10px;
  text-align: center;
  margin-bottom: 20px;
  color: #fff;
}

.box.selected {
  border-color: greenyellow;
}

.arrow_box:not(:last-of-type):after, .arrow_box:not(:last-of-type):before {
  top: 100%;
  left: 50%;
  border: solid transparent;
  content: " ";
  height: 0;
  width: 0;
  position: absolute;
  pointer-events: none;
}

.arrow_box:not(:last-of-type):after {
  border-color: rgba(136, 183, 213, 0);
  border-top-color: #88b7d5;
  border-width: 20px;
  margin-left: -20px;
}

.arrow_box:not(:last-of-type):before {
  border-color: rgba(194, 225, 245, 0);
  border-top-color: #c2e1f5;
  border-width: 24px;
  margin-left: -24px;
}

.ui.mini.button.move-btn {
  font-weight: 600;
  margin-bottom: 5px;
}

.ui.grid.group-requests {
  padding-bottom: 25px;
}