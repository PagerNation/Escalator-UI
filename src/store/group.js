import 'whatwg-fetch';
import _ from 'lodash';
import { getJSON, deleteObject, putJSON } from "../utils/apiRequest";

// ------------------------------------
// Constants
// ------------------------------------
export const FETCH_GROUP_SUCCESS = 'FETCH_GROUP_SUCCESS';
export const LEAVE_GROUP_SUCCESS = 'LEAVE_GROUP_SUCCESS';
export const UPDATE_EP_SUCCESS = 'UPDATE_EP_SUCCESS';

// ------------------------------------
// Actions
// ------------------------------------
export const fetchGroup = (groupId) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      getJSON('group/' + groupId).then((response) => {
        dispatch({
          type: FETCH_GROUP_SUCCESS,
          payload: response
        });
        resolve();
      });
    });
  }
};

export const leaveGroup = (groupName, userId) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      deleteObject(`group/${groupName}/user/${userId}`).then((json) => {
        dispatch({
          type: LEAVE_GROUP_SUCCESS,
          payload: json
        });
        resolve();
      });
    })
  }
};

export const updateEscalationPolicy = (groupName, ep) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      putJSON(`group/${groupName}/escalationpolicy`, ep).then((response) => {
        dispatch({
          type: UPDATE_EP_SUCCESS,
          payload: response
        });
        resolve();
      });
    })
  }
};


// ------------------------------------
// Action Handlers
// ------------------------------------
const ACTION_HANDLERS = {
  [FETCH_GROUP_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [LEAVE_GROUP_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [UPDATE_EP_SUCCESS]: (state, action) => {
    return action.payload;
  }
};

// ------------------------------------
// Reducer
// ------------------------------------
const initialState = null;

export default function groupReducer(state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type];
  return handler ? handler(state, action) : state;
};
