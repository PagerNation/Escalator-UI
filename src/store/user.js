import 'whatwg-fetch';
import _ from 'lodash';
import jwtDecode from 'jwt-decode';
import { getJSON, postJSON, putJSON, deleteObject } from '../utils/apiRequest';
import { LOG_IN_FAILURE, SIGN_UP_FAILURE } from "./api";

// ------------------------------------
// Constants
// ------------------------------------
export const LOG_IN_SUCCESS = 'LOG_IN_SUCCESS';
export const LOG_OUT = 'LOG_OUT';

export const FETCH_USER_SUCCESS = 'FETCH_USER_SUCCESS';
export const ADD_DEVICE_SUCCESS = 'ADD_DEVICE';
export const DELETE_DEVICE_SUCCESS = 'DELETE_DEVICE_SUCCESS';
export const FETCH_GROUPS_SUCCESS = 'FETCH_GROUPS_SUCCESS';
export const UPDATE_PROFILE_SUCCESS = 'UPDATE_PROFILE_SUCCESS';
export const REORDER_DEVICES_SUCCESS = "REORDER_DEVICES_SUCCESS";


// ------------------------------------
// Actions
// ------------------------------------
export const logIn = (email, password) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      postJSON('auth/login', {email, password}).then((response) => {
        dispatch({
          type: LOG_IN_SUCCESS,
          payload: response
        });
        resolve();
      }).catch((response) => {
        dispatch({
          type: LOG_IN_FAILURE,
          payload: response
        });
        resolve();
      });
    });
  }
};

export const signUp = (name, email, password, passwordConfirm) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      if (password !== passwordConfirm) {
        dispatch({
          type: SIGN_UP_FAILURE,
          payload: "Passwords do not match."
        });
      } else {
        postJSON('auth/signup', {name, email, password}).then((response) => {
          dispatch({
            type: LOG_IN_SUCCESS,
            payload: response
          });
          resolve();
        }).catch((response) => {
          dispatch({
            type: SIGN_UP_FAILURE,
            payload: response
          });
          resolve();
        });
      }
    });
  }
};

export const logOut = () => {
  return (dispatch, getState) => {
    dispatch({
      type: LOG_OUT
    });
  }
};

export const fetchUser = () => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const userId = jwtDecode(localStorage.getItem('escalatorToken')).id;
      getJSON(`user/${userId}/`).then((response) => {
        dispatch({
          type: FETCH_USER_SUCCESS,
          payload: response
        });
        resolve();
      });
    });
  }
};

export const updateProfile = (profile) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const userId = jwtDecode(localStorage.getItem('escalatorToken')).id;
      putJSON(`user/${userId}/`, profile).then((response) => {
        dispatch({
          type: UPDATE_PROFILE_SUCCESS,
          payload: response
        });
        resolve();
      });
    });
  }
};

export const addDevice = (device) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const userId = jwtDecode(localStorage.getItem('escalatorToken')).id;
      postJSON(`user/${userId}/device`, device).then((response) => {
        dispatch({
          type: ADD_DEVICE_SUCCESS,
          payload: response
        });
        resolve();
      });
    })
  }
};

export const deleteDevice = (id) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const userId = jwtDecode(localStorage.getItem('escalatorToken')).id;
      deleteObject(`user/${userId}/device/${id}`).then((json) => {
        dispatch({
          type: DELETE_DEVICE_SUCCESS,
          payload: json
        });
        resolve();
      });
    })
  }
};

export const reorderDevices = (ids) => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const userId = jwtDecode(localStorage.getItem('escalatorToken')).id;
      putJSON(`user/${userId}/devices/`, {sortOrder: ids}).then((response) => {
        dispatch({
          type: REORDER_DEVICES_SUCCESS,
          payload: response
        });
        resolve();
      });
    })
  }
};

// ------------------------------------
// Action Handlers
// ------------------------------------
const ACTION_HANDLERS = {
  [LOG_IN_SUCCESS]: (state, action) => {
    localStorage.setItem('escalatorToken', action.payload.token);
    return action.payload.user;
  },
  [LOG_OUT]: (state, action) => {
    localStorage.removeItem('escalatorToken');
    return null;
  },
  [FETCH_USER_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [ADD_DEVICE_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [DELETE_DEVICE_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [REORDER_DEVICES_SUCCESS]: (state, action) => {
    return action.payload;
  },
  [FETCH_GROUPS_SUCCESS]: (state, action) => {
    const newState = _.extend({}, state);
    newState.groups = action.payload.groups;
    return newState;
  },
  [UPDATE_PROFILE_SUCCESS]: (state, action) => {
    return action.payload;
  }
};

// ------------------------------------
// Reducer
// ------------------------------------
const initialState = null;

export default function userReducer(state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type];
  return handler ? handler(state, action) : state;
};
