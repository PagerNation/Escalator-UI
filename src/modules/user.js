import { injectReducer } from '../store/reducers';

// ------------------------------------
// Constants
// ------------------------------------
export const FETCH_USER = 'FETCH_USER';

// ------------------------------------
// Actions
// ------------------------------------
export function fetchUser() {
  return {
    type: FETCH_USER,
    payload: value
  }
}

export const actions = {
  fetchUser
};

// ------------------------------------
// Action Handlers
// ------------------------------------
const ACTION_HANDLERS = {
  [FETCH_USER]: (state, action) => {
    // TODO: Use endpoint
    return {
      id: '693d9a0698aff95c',
      name: 'Kevin Moses',
      email: 'kwm4385@rit.edu',
      groups: [{
        type: "Group",
        ref: 'Some group'
      },
        {
          type: "Group",
          ref: 'Awesome Group'
        }],
      devices: [{
        name: 'Cell',
        type: 'sms',
        contactInformation: '585-328-5354',
        createdAt: Date.now()
      }]
    };
  }
};

// ------------------------------------
// Reducer
// ------------------------------------
const initialState = null;

export default function userReducer(state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type];

  return handler ? handler(state, action) : state;
};

injectReducer(store, {
  key: 'user',
  reducer: userReducer
});
