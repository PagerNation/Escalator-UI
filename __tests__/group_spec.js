import React from 'react';
import GroupView from '../src/routes/Group/components/GroupView';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import group from './fixtures/group_response.json';
import user from './fixtures/user.json';
import tickets from './fixtures/tickets.json'

describe('GroupView', () => {
  it('renders GroupView', () => {
    spyOn(GroupView.prototype, 'componentWillMount').and.returnValue(null);
    const component = renderer.create(<GroupView group={group} user={user} tickets={tickets}/>);
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  })
});
