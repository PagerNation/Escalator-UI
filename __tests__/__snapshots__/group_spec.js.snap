// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupView renders GroupView 1`] = `
<div>
  <h1
    className="ui header"
  >
    AudraDickinson
  </h1>
  <button
    className="ui button action-button"
    onClick={[Function]}
    tabIndex={undefined}
  >
    Leave Group
  </button>
  <h4
    className="ui header"
  >
    User on call:
    <a
      href="/user/58a09497b73f14c7d6d64437/"
    >
      Kelly85
    </a>
  </h4>
  <span>
    Escalation Interval: 
    <div
      className="ui teal horizontal label"
      onClick={[Function]}
    >
      10
    </div>
     minutes
  </span>
  <div
    className="ui divider"
  />
  <div
    className="ui grid"
  >
    <div
      className="seven wide computer sixteen wide mobile column"
    >
      <h3>
        Escalation Order:
      </h3>
      <a
        className="box arrow_box"
        onClick={[Function]}
      >
        Kelly85
      </a>
      <a
        className="box arrow_box"
        onClick={[Function]}
      >
        Demetrius.Hahn
      </a>
    </div>
    <div
      className="middle aligned one wide computer sixteen wide mobile column"
    />
    <div
      className="seven wide computer sixteen wide mobile column"
    >
      <h3>
        Not On Call:
      </h3>
      <a
        className="box"
        onClick={[Function]}
      >
        Seth4
      </a>
    </div>
  </div>
</div>
`;
